src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java

```diff
@@ -28,56 +28,191 @@
+ * any other context.  This implies that a constructor may choose to
+ * initialize a stable variable, rather than "leaving it for later",
+ * and that initial will be safely published, as if the field were
+ * {@code final}.  The stored value may (or may not) contain
+ * additional stable variables, not yet initialized.  Note that if a
+ * stable variable is written outside of the code of a constructor (or
+ * class initializer), then data races are possible, just the same as
+ * if there were no {@code @Stable} annotation, and the field was a
+ * regular mutable field.  In fact, the usual case for lazily
+ * evaluated data structures is to assign to stable variables much
+ * later than the enclosing data structure is created.  This means
+ * that racing reads and writes might observe nulls (or primitive
+ * zeroes) as well as non-default values.
+ * <p>
+ * Therefore, most code which reads stable variables should not assume
+ * that the value has been set, and should dynamically test for a null
+ * (or zero) value.  Code which cannot prove proper ordering of
+ * initialization may use stable variables without performing the null
+ * (or zero) test.  Code which omits the null (or zero) test should be
+ * documented as to why the initialization order is reliable.  In
+ * general, some sort of critical section for initialization should be
+ * pointed out as provably preceding all uses of the (unchecked)
+ * stable variable.  Examples of such a critical section would be a
+ * constructor body which directly writes a final or stable variable,
+ * or the AOT cache assembly phase as a whole.
+ * <p>
+ * The HotSpot VM relies on this annotation to promote a non-null (or
+ * non-zero) stable variable use to a constant, thereby enabling superior
+ * optimizations of code depending on the value (such as constant folding).
  * More specifically, the HotSpot VM will process non-null stable fields (final
  * or otherwise) in a similar manner to static final fields with respect to
  * promoting the field's value to a constant.  Thus, placing aside the
  * differences for null/non-null values and arrays, a final stable field is
  * treated as if it is really final from both the Java language and the HotSpot
  * VM.
  * <p>
- * It is (currently) undefined what happens if a field annotated as stable
- * is given a third value (by explicitly updating a stable field, a component of
- * a stable array, or a final stable field via reflection or other means).
- * Since the HotSpot VM promotes a non-null component value to constant, it may
- * be that the Java memory model would appear to be broken, if such a constant
- * (the second value of the field) is used as the value of the field even after
- * the field value has changed (to a third value).
+ * After constant folding, the compiler can make use of may aspects of
+ * the object: Its dynamic type, its length (if it is an array), and
+ * the values of its fields (if they are themselves constants, either
```

Comment:
I wonder if this includes object identity, or the object's identity hash code - for example, published MethodType objects are all identity-based, and using their identity hash should be better than using their nominal hash code.